### PR TITLE
resolves #32

### LINKED ISSUE #32: Handle extreme edge cases in production environments
The problem is with extreme rare edge cases when using [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#train-test-split) without specifying `stratify` parameter. The following mismatches were spotted and should be somehow resolved:
- how binary would behave when seeing new element in X_test
- how nominal and ordinal would behave in X_test in the same scenario

Possible solution:
- add `binary_value_order` to config as optional parameter, if omitted we auto detect unique values
- add to config how to handle first seen parameter in cat_nominal
- the same as above for ordinal

One should add separate `production` section to documentation, describing these changes and specific scenarios.

### PR DESCRIPTION
**Pull request by BartekKruczek**
- resolves #32 
- introduce more Python performance adjustments
- fix typo issues

### CODE REVIEW

File: gower_metric/utils/transforms/categorical.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        encoded = (
+            enc.transform(col_arr[non_null_mask].reshape(-1, 1))
+            .astype(data_type)
+            .ravel()
+        )
+        transformed_col[non_null_mask] = encoded
+
+    if handle_unseen == "warning":
+        nan_output = np.isnan(transformed_col)
+        unseen_mask = non_null_mask & nan_output
```
[the remark this answers is not in the document]
Comment (BartekKruczek) [reply gower_metric/utils/transforms/categorical.py:42]: Not valid after np.int dtype drop

File: docs/source/modules/user_guide/config.rst [hunk 2]
```diff
@@ -46,4 +48,50 @@ some of these configuration flags depend on one another. It will be clearly indi
...
+- ``"error"`` (default) – raise a ``ValueError`` when an unseen value is encountered. This is the safest option and forces you to deal with the issue explicitly.
+- ``"warning"`` – silently map the unseen value to ``NaN`` (i.e. treat it as missing), but emit a ``UserWarning`` so you know it happened.
+- ``"missing"`` – silently map it to ``NaN`` without any warning. Handy when you already know unseen values are expected and acceptable.
+
+Each feature type has its own toggle:
+
+- ``handle_unseen_binary_asymmetric (str)`` - Strategy for handling unseen values in binary asymmetric features. Defaults to ``"error"``.
+- ``handle_unseen_binary_symmetric (str)`` - Strategy for handling unseen values in binary symmetric features. Defaults to ``"error"``.
+- ``handle_unseen_categorical_nominal (str)`` - Strategy for handling unseen values in categorical nominal features. Defaults to ``"error"``.
+- ``handle_unseen_categorical_ordinal (str)`` - Strategy for handling unseen values in categorical ordinal features. Defaults to ``"error"``.
```
Comment (mwardynski) [inline docs/source/modules/user_guide/config.rst:72]: what will happen to ratio scale/numeric interval values which exceed max, or are smaller than min? Don't we calculate these values during the fit?

What if `iqr` or `kde` are turned on.

File: gower_metric/core/config.py [hunk 3]
```diff
@@ -118,15 +145,15 @@ def check_scale_window_type(
...
             return v
 
         scale_window = info.data.get("scale_window")
         if scale_window is None:
-            if v is not None:
+            if v is not None:  # pragma: no branch
                 msg = "scale_window_type must be None when scale_window is None"
                 raise ValueError(msg)
-        elif scale_window == "kde" and v not in (None, "silverman"):
+        elif scale_window == "kde" and v not in (None, "silverman"):  # pragma: no cover
```
Comment (mwardynski) [inline gower_metric/core/config.py:165]: do we want to introduce pragma to skip  lines calculating the coverage?

File: gower_metric/core/config.py [hunk 4]
```diff
@@ -251,3 +278,233 @@ def check_threshold_coeff(
             msg = f"conditional_distances_threshold_coeff must be at least 1, got {v}"
             raise ValueError(msg)
         return v
+
+    @field_validator("handle_unseen_binary_asymmetric")
+    @classmethod
+    def check_handle_unseen_binary_asymmetric(
```
Comment (mwardynski) [inline gower_metric/core/config.py]: we don't need an explicit validator, if acceptable values are defined in `Literal`, do we?

Comment (mwardynski) [reply gower_metric/core/config.py]: the same comment apply to many of  the validators below

File: gower_metric/core/config.py [hunk 5]
```diff
@@ -251,3 +278,233 @@ def check_threshold_coeff(
...
+        valid_strategies = get_args(HandleUnseenBinarySymmetric)
+        if v not in valid_strategies:  # pragma: no cover
+            msg = f"handle_unseen_binary_symmetric must be one of {valid_strategies}, got {v}"
+            raise ValueError(msg)
+
+        return v
+
+    @field_validator("binary_symmetric_value_order")
+    @classmethod
+    def check_binary_symmetric_value_order(
```
Comment (mwardynski) [inline gower_metric/core/config.py:369]: it's repeating exactly the same code like for the asymmetric value order - either use the same function with multiple `@field_validator` to validate both, or extract the common part to an additional function

File: gower_metric/core/metric.py [hunk 6]
```diff
@@ -281,14 +310,26 @@ def fit(self, X: pd.DataFrame | np.ndarray) -> "Gower":
         self.cat_ord_metadata: dict[int | str, dict[str, Any]] = {}
         for j in self.categorical_ordinal_indices:
             col = arr[:, j]
-            if self.categorical_ordinal_values_order is None:
+            if self.categorical_ordinal_values_order is None:  # pragma: no cover
```
Comment (mwardynski) [inline gower_metric/core/metric.py]: I'm not sure whether introducing `#pragma` for excluding from test coverage was a good idea...
e.g. why the line above has been excluded? Can't `categorical_ordinal_values_order` be `None` in this place?

File: gower_metric/core/metric.py [hunk 7]
```diff
@@ -281,14 +310,26 @@ def fit(self, X: pd.DataFrame | np.ndarray) -> "Gower":
...
+            counts_arr = np.asarray(
+                [
+                    counts_map.get(v, 0)
+                    for v in self.categorical_ordinal_values_order[j]
+                ],
+                dtype=float,
+            )
+
+            # we need to map ranks to both int and float to avoid cast errors within .fit()
+            # to maintain compatibility with multiple .transform() calls
```
Comment (mwardynski) [inline gower_metric/core/metric.py]: can you explain this case?

File: gower_metric/utils/matrix/convert_matrix.py [hunk 8]
```diff
@@ -56,14 +58,14 @@ def __get_coo_matrix(
 def get_scipy_sparse_matrix(
```
Comment (mwardynski) [inline gower_metric/utils/matrix/convert_matrix.py:58]: aren't we refusing the processing of a sparse matrix? why do we need this function?

File: gower_metric/utils/transforms/binary.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+                "Consider using binary_asymmetric_value_order to explicitly define the expected binary values, "
+                "or change the feature type if this is not actually a binary feature."
+            )
+            raise ValueError(msg)
+
+    for i, v in enumerate(col):
+        if pd.isna(v):
+            transformed_col[i] = np.nan
+        elif v not in mapping:
+            if is_explicit:
```
Comment (mwardynski) [inline gower_metric/utils/transforms/binary.py:67]: do we really need 3 variants of this `msg`?

Comment (mwardynski) [reply gower_metric/utils/transforms/binary.py:67]: the same comment goes to many other functions below

File: tests/gower_api/precision/test_binary_symmetric_value_order.py
Comment (mwardynski) [inline tests/gower_api/precision/test_binary_symmetric_value_order.py:1]: Isn't this file like in 80% the same, as for. binary asymmetric? If so, how about we reuse some code here and there?

File: tests/gower_api/unit/test_matrix_coverage.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        data = pd.DataFrame({"x": [1.0, 5.0, 10.0]})
+        cfg = Config(feature_types={"x": "numeric"}, data_type=np.float64)
+        gower = Gower(cfg).fit(data)
+        mat = gower.matrix(data)
+        assert mat.dtype == np.float64
+
+
+class TestSparseFloat16Upcast:
+    def test_float16_upcasts_to_float32_with_warning(self) -> None:
+        """scipy.sparse does not support float16; verify upcast + warning."""
```
Comment (mwardynski) [inline tests/gower_api/unit/test_matrix_coverage.py:86]: again sparse, aren't they throw an error in the first place?

File: tests/gower_api/unit/test_transform_branches.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            feature_types={"flag": "binary_asymmetric", "value": "numeric"},
+            binary_asymmetric_value_order={"flag": [0, 1]},
+        )
+        gower = Gower(cfg).fit(train)
+        with pytest.raises(ValueError, match="violates binary_asymmetric_value_order"):
+            gower.transform(test)
+
+
+class TestCategoricalUnseenWarningBranch:
+    def test_nominal_unseen_warning(self) -> None:
```
Comment (mwardynski) [inline tests/gower_api/unit/test_transform_branches.py]: similar tests are already here `tests/gower_api/precision/test_handle_unseen_categorical_nominal.py` why do we repeat them? The same comment applies to many other test types

File: MANIFEST.in [hunk 12]
```diff
@@ -2,6 +2,7 @@ include README.md
 include LICENSE.txt
 include pyproject.toml
 include uv.lock
+include requirements.txt
```
Comment (mwardynski) [inline MANIFEST.in:5]: why do we need `requirements.txt`, when we use `uv` with `pyproject.toml`?

File: docs/source/modules/user_guide/config.rst (hunk 2 above)
Comment (BartekKruczek) [reply docs/source/modules/user_guide/config.rst:72]: Good point!

File: gower_metric/core/config.py (hunk 3 above)
Comment (BartekKruczek) [reply gower_metric/core/config.py:165]: We don't, will delete it later

File: MANIFEST.in (hunk 12 above)
Comment (BartekKruczek) [reply MANIFEST.in:5]: Many people still prefer to use `pip` :/

File: gower_metric/core/config.py (hunk 4 above)
Comment (BartekKruczek) [reply gower_metric/core/config.py]: Correct, user can not pass something that is not within `Literal`. Will delete these validators

File: gower_metric/core/config.py (hunk 5 above)
Comment (BartekKruczek) [reply gower_metric/core/config.py:369]: Done

File: gower_metric/core/metric.py (hunk 6 above)
Comment (BartekKruczek) [reply gower_metric/core/metric.py]: Well, can be. `Pragma` deleted

File: gower_metric/utils/matrix/convert_matrix.py (hunk 8 above)
Comment (BartekKruczek) [reply gower_metric/utils/matrix/convert_matrix.py:58]: We do not take sparse matrix as input. However one can convert and get output as `scipy.sparse.matrix`.

User calls `matrix(convert_to_sparse=True)` -> calculations -> scipy sparse matrix as direct output

File: gower_metric/utils/transforms/binary.py (hunk 9 above)
Comment (BartekKruczek) [reply gower_metric/utils/transforms/binary.py:67]: We don't... will simplify it

File: gower_metric/core/metric.py (hunk 7 above)
Comment (BartekKruczek) [reply gower_metric/core/metric.py]: Sure. `map_ordered_values` returns something like `{"low": 0, "medium": 1, "high": 2}`, then OrdinalEncoder encodes columns based on passed `data_type`, here into floats. `map_ordered_values` uses `int` dtype, thus creating difference between value's types. Then, we use `r_x = np.array([ranks_map.get(v, np.nan) for v in col_x])` which searches for mapped (converted) columns. 

Consider it as fail safe to include both float and int dtypes into mapping thus enabling multiple `transform()` calls without messing with lookups. What is more, it will enable in the future `int` datatype as input

File: tests/gower_api/unit/test_transform_branches.py (hunk 11 above)
Comment (BartekKruczek) [reply tests/gower_api/unit/test_transform_branches.py]: Will take closer look

File: tests/gower_api/precision/test_binary_symmetric_value_order.py
Comment (BartekKruczek) [reply tests/gower_api/precision/test_binary_symmetric_value_order.py:1]: Done

File: tests/gower_api/unit/test_matrix_coverage.py (hunk 10 above)
Comment (BartekKruczek) [reply tests/gower_api/unit/test_matrix_coverage.py:86]: Nope, `float16` + dense matrix as input is acceptable. However, scipy does not support [convertion](https://github.com/scipy/scipy/blob/main/scipy/sparse/_sputils.py) from float16 -> thus upscaling to float32

File: tests/gower_api/unit/test_transform_branches.py (hunk 11 above)
Comment (BartekKruczek) [reply tests/gower_api/unit/test_transform_branches.py]: Done

File: docs/source/modules/user_guide/config.rst (hunk 2 above)
Comment (BartekKruczek) [reply docs/source/modules/user_guide/config.rst:72]: Done. Added:
- new flag to Config; enables either silent clipping ([0, 1]), clipping with user warning or pure error rise when out of range (oor) value(s) detected
- new functions for detecting and handling oor (gower/utils/ranges)
- adjusted main source file

File: docs/source/modules/installation/contribution.rst [hunk 13]
```diff
@@ -2,4 +2,4 @@
 Contribution
 =============
 
-For contributions, please refer to the `CONTRIBUTING.md <https://github.com/mwardynski/gower-metric/blob/refactor_newapi/.github/CONTRIBUTING.md>`_ file.
+For contributions, please refer to the `CONTRIBUTING.md <https://github.com/Gower-metric/gower-metric/blob/master/.github/CONTRIBUTING.md>`_ file.
```
Comment (mwardynski) [inline docs/source/modules/installation/contribution.rst]: we should use for the branch name: "main" instead of "master"

File: docs/source/modules/user_guide/transform.rst [hunk 14]
```diff
@@ -58,4 +58,84 @@ For convenience, Gower also implements ``fit_transform`` method, which combines
...
+
+   import numpy as np
+   from gower_metric import Config, Gower
+
+   X_train = np.array([[1.0], [5.0]], dtype=object)
+   X_test = np.array([[10.0]], dtype=object)  # outside [1, 5]
+
+   cfg = Config(
+       feature_types={0: "numeric"},
+       out_of_range="error",  # raise instead of warn
```
Comment (mwardynski) [inline docs/source/modules/user_guide/transform.rst]: the comment is missing the word "error", but anyway when we change the default behavior to "error", we may need to change this example and the comment.

File: docs/source/modules/user_guide/transform.rst [hunk 15]
```diff
@@ -58,4 +58,84 @@ For convenience, Gower also implements ``fit_transform`` method, which combines
...
\ No newline at end of file
+   transformed_data = gower.fit_transform(data)
+
+-----------------------------------
+Out-of-range values in transform
+-----------------------------------
+
+When numeric or ratio-scale features in the transform (or test) data contain values outside the range
+observed during ``fit()``, the ``out_of_range`` config parameter controls the behavior.
+By default it is set to ``"warning"`` — a ``UserWarning`` is emitted listing each column where values
```
Comment (mwardynski) [inline docs/source/modules/user_guide/transform.rst]: we should set it by default to "error" so it stays consistent with unseen values.

File: gower_metric/core/config.py [hunk 16]
```diff
@@ -72,6 +96,13 @@ class Config(BaseModel):
     k_neighbors: int | None = None
     conditional_distances: ConditionalDistancesFlag = False
     conditional_distances_threshold_coeff: int = 1
+    handle_unseen_binary_asymmetric: HandleUnseenBinaryAsymmetric = "error"
+    binary_asymmetric_value_order: BinaryAsymmetricValueOrderType = None
+    handle_unseen_binary_symmetric: HandleUnseenBinarySymmetric = "error"
+    binary_symmetric_value_order: BinarySymmetricValueOrderType = None
+    handle_unseen_categorical_nominal: HandleUnseenCategoricalNominal = "error"
+    handle_unseen_categorical_ordinal: HandleUnseenCategoricalOrdinal = "error"
+    out_of_range: OutOfRangeStrategy = "warning"
```
Comment (mwardynski) [inline gower_metric/core/config.py]: "error" ?

File: gower_metric/core/metric.py (hunk 7 above)
Comment (mwardynski) [reply gower_metric/core/metric.py]: wouldn't the right solution be to make `map_ordered_values` return mappings, which already respect set `data_type` in the processing?

File: gower_metric/core/metric.py [hunk 17]
```diff
@@ -382,80 +464,69 @@ def transform(self, X: pd.DataFrame | np.ndarray) -> pd.DataFrame | np.ndarray:
             arr: np.ndarray = X
             X_arr = np.asarray(arr, dtype=object)
 
+        enforce_oor_policy(
```
Comment (mwardynski) [inline gower_metric/core/metric.py]: shouldn't the call of this function take into account the value of `_skip_oor_check`?

File: gower_metric/core/metric.py [hunk 18]
```diff
@@ -771,22 +854,47 @@ def matrix(
...
-            matrix_type=matrix_type,
-            convert_to_sparse=convert_to_sparse,
-            sparse_type=sparse_type,
-            backend=backend,
+        arr_check = (
+            X.to_numpy(dtype=object)
+            if isinstance(X, pd.DataFrame)
+            else np.asarray(X, dtype=object)
         )
+        enforce_oor_policy(
```
Comment (mwardynski) [inline gower_metric/core/metric.py]: what if `_skip_oor_check` is set to true?

File: gower_metric/core/metric.py [hunk 19]
```diff
@@ -771,22 +854,47 @@ def matrix(
...
+            strategy=self.out_of_range,
+            numeric_indices=self.numeric_indices,
+            numeric_mins=self.numeric_mins,
+            numeric_maxs=self.numeric_maxs,
+            ratio_scale_indices=self.ratio_scale_indices,
+            ratio_mins=self.ratio_mins,
+            ratio_maxs=self.ratio_maxs,
+            stacklevel=2,
+        )
+        self._skip_oor_check = True
```
Comment (mwardynski) [inline gower_metric/core/metric.py]: Overriding the user's settings seem to be a bit sketchy.
Also the user may override once set values and wouldn't even know, how and why

File: MANIFEST.in (hunk 12 above)
Comment (mwardynski) [reply MANIFEST.in:5]: and they still can freely use our lib once the wheel is published to PyPI with pip or any tool they want.
It doesn't mean, that we need to provide means to work with the project on the code level with pip - providing `pyproject.toml` should be enough

File: docs/source/modules/user_guide/transform.rst (hunk 15 above)
Comment (BartekKruczek) [reply docs/source/modules/user_guide/transform.rst]: ok

File: docs/source/modules/user_guide/config.rst [hunk 20]
```diff
@@ -46,4 +48,99 @@ some of these configuration flags depend on one another. It will be clearly indi
...
-  then the final distance will be set to 1. Otherwise, the ratio scale and numeric features will be included in the distance calculation in the second step.
\ No newline at end of file
+  then the final distance will be set to 1. Otherwise, the ratio scale and numeric features will be included in the distance calculation in the second step.
+- ``out_of_range (str)`` - Strategy for handling numeric and ratio-scale values that fall outside the range observed during ``fit()``. Possible values are ``"clip"`` (silently clip normalized distances to [0, 1]), ``"warning"`` (default, emit a ``UserWarning`` listing offending columns and clip), or ``"error"`` (raise a ``ValueError``). This applies to both ``transform()`` and pairwise distance calculations (``__call__``, ``matrix``).
+- ``skip_out_of_range_validation (StrictBool)`` - Determine whether to skip validation of out-of-range values. Settings to ``False`` might
+    increase overall speedup, however it is not recommended without great care and solid data preprocessing. Default to ``False`` if omitted.
```
Comment (mwardynski) [inline docs/source/modules/user_guide/config.rst]: I thiink it should be: "Settings to ``True`` might increase..."
Besides please don't use the word speedup - most computer scientist associate speedup as a metric for parallel computing. Rephrase it a bit

**Review by mwardynski — APPROVED** [review]
